pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,244 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from copy import deepcopy
+from typing import Any, Dict, Generator, Iterator, List, Optional, Union
+
+from torch.utils.data import BatchSampler, get_worker_info, Sampler
+from torch.utils.data.dataloader import IterableDataset
+
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler(Sampler):
```

Comment:
I'm not sure if this should subclass `Sampler`. This is a "has a" relationship, not a "is a".

For example, in a similar wrapper, we don't subclass:

https://github.com/PyTorchLightning/pytorch-lightning/blob/01f594baf4d2d66515b6f330dee017e8a138ee9c/pytorch_lightning/overrides/distributed.py#L126

cc @justusschock who asked for this change

Reply:
I think it is safer to be a Sampler. It doesn't have to be, but users might have to assert checking a sampler is provided.